core/internal/filetransfer/file_transfer_default.go

```diff
@@ -69,10 +70,22 @@ func (ft *DefaultFileTransfer) Upload(task *Task) error {
 		)
 	}
 
-	task.Size = stat.Size()
+	if task.Offset > 0 || (task.Length > 0 && task.Length < stat.Size()) {
+		if task.Offset+task.Length > stat.Size() {
+			ft.logger.Error("file transfer: upload: offset + length is greater than the file size",
+				"offset", task.Offset, "length", task.Length, "file size", stat.Size())
+			task.Length = int64(stat.Size() - task.Offset)
+		}
+		sectionReader := io.NewSectionReader(file, task.Offset, task.Length)
+		reader = sectionReader
+		task.Size = task.Length
```

Comment:
i think in any case this requires explanation it's not clear why you do the math you do
why it doesn't include the range